crates/shared/src/gas_price_estimation/driver.rs

```diff
@@ -0,0 +1,119 @@
+use {
+    anyhow::{Context, Result},
+    gas_estimation::{GasPrice1559, GasPriceEstimating},
+    number::serialization::HexOrDecimalU256,
+    primitive_types::U256,
+    reqwest::Url,
+    serde::Deserialize,
+    serde_with::serde_as,
+    std::{
+        sync::{Arc, Mutex},
+        time::{Duration, Instant},
+    },
+    tracing::instrument,
+};
+
+/// A gas price estimator that fetches gas prices from a driver instance with
+/// caching.
+#[derive(Clone)]
+pub struct DriverGasEstimator {
+    client: reqwest::Client,
+    url: Url,
+    cache: Arc<Mutex<Option<CachedGasPrice>>>,
```

Comment:
Q: Is this not a `std::sync::RwLock` to avoid the issues they mention in the docs (writer starvation and possible deadlocks=?

Reply:
I actually didn't think much about this choice to be honest. I feel since there is not a whole lot of uses for concurrent reads here `RwLock` may be overkill?

Reply:
I don't have an idea on numbers (reads/writes) so that might be ok. What prompted me asking was the fine-grained locking under [`estimate()`](https://github.com/cowprotocol/services/pull/3569/files#diff-ebe27455bc6fade9ebb5ba68a4c9f9e531e961b1b3213a7e4523d36e544fe93aR82-R91) which (at first glance) seemed that it could benefit from it since the first read is locked and multiple readers would block there.

But again, I don't have a good grasp of the system yet, so this was more of a probing question

Reply:
Increased the critical section to include write, which makes more sense and the choice of mutex obvious.